File: apps/web/src/components/playout/playhead.ts
import { useEffect, useState } from 'react';
import type { PlayoutNowPlaying } from '@deadair/sdk';

/** How often the local clock moves between readings. Fast enough to look continuous, slow enough to be free. */
const TICK_MS = 500;

export interface Playhead {
    /** Milliseconds into the item on air. */
    elapsedMs: number;
    /** Milliseconds left of it. */
    remainingMs: number;
    /** Its full length, which is what makes the other two meaningful. */
    durationMs: number;
    /** 0–100, for a progress bar. */
    percent: number;
}

/**
 * The playhead of the item on air, projected between readings.
 *
 * The API answers with the DECODER's own countdown, taken at the moment it was
 * read, and the console polls every couple of seconds. Showing that number raw
 * gives a clock that jumps two seconds at a time; projecting from it gives one
 * that moves, and re-anchors on every reading rather than drifting away from the
 * station. It is the same projection the API's own rundown does when it answers,
 * for the same reason.
 *
 * Returns `undefined` when there is nothing to measure — no item, no reported
 * duration, or a decoder that could not say how much is left. A clock
 * extrapolated from a start time would be a moving, confident lie, and that is
 * exactly the thing the transport is not allowed to be.
 */
export function usePlayhead(nowPlaying: PlayoutNowPlaying | undefined): Playhead | undefined {
    const itemId = nowPlaying?.item.id;
    const durationMs = nowPlaying?.item.durationMs;
    const reportedMs = nowPlaying?.remainingMs;
    const measured = durationMs !== undefined && reportedMs !== undefined && durationMs > 0;

    // How far the clock has been carried past the last reading. Reset during render
    // rather than from an effect, because it is derived from the props: an effect
    // would render one frame of the previous track's clock first.
    //
    // Counted in ticks rather than measured against a wall clock, which keeps this
    // a pure render. It also fails in the safer direction: a throttled background
    // tab under-counts, so the clock stalls until the next reading re-anchors it,
    // where a wall clock would jump.
    const reading = `${itemId ?? ''}:${reportedMs ?? ''}`;
    const [carried, setCarried] = useState({ reading, ms: 0 });
    if (carried.reading !== reading) setCarried({ reading, ms: 0 });

    useEffect(() => {
        if (!measured) return;

        // `reading` is a dependency so the interval restarts with each one, and a
        // tick always measures a full interval since the station last spoke.
        const timer = setInterval(() => {
            setCarried(current => ({ ...current, ms: current.ms + TICK_MS }));
        }, TICK_MS);
        return () => {
            clearInterval(timer);
        };
    }, [measured, reading]);

    if (!measured) return undefined;

    const remainingMs = Math.min(durationMs, Math.max(0, reportedMs - carried.ms));
    const elapsedMs = durationMs - remainingMs;
    return { elapsedMs, remainingMs, durationMs, percent: (elapsedMs / durationMs) * 100 };
}

File: apps/desktop/src/MaroonedSoftware.Deadair.Desktop.Core/NowPlaying/Playhead.cs
using MaroonedSoftware.Deadair.Sdk.Models;

namespace MaroonedSoftware.Deadair.Desktop.Core.NowPlaying;

/// <summary>
/// Where the record is up to, between readings.
/// </summary>
/// <remarks>
/// <para>
/// The station is polled every few seconds and a playhead moves every frame, so the position shown
/// is projected locally from the last reading and re-anchored whenever a new one arrives.
/// </para>
/// <para>
/// <b>It refuses to answer when the station could not say.</b> <c>remainingMs</c> is absent when the
/// decoder does not know, and the tempting fallback — subtracting <c>startedAt</c> from the clock —
/// is wrong in a way that looks right: it measures how long ago the station STARTED the record,
/// which leads what a listener is hearing by the encoder and buffer, and drifts further the worse
/// somebody's connection is. A progress bar that is confidently wrong is worse than one that is
/// absent, so this returns null and the view draws nothing.
/// </para>
/// </remarks>
public static class Playhead
{
    /// <summary>
    /// How far into the record the listener is, or null when it cannot be said.
    /// </summary>
    /// <param name="track">The track from the last reading.</param>
    /// <param name="readAt">When that reading arrived.</param>
    /// <param name="now">The current time.</param>
    public static TimeSpan? Position(NowPlayingTrack? track, DateTimeOffset? readAt, DateTimeOffset now)
    {
        if (track?.DurationMs is not { } duration || duration <= 0)
        {
            return null;
        }

        if (track.RemainingMs is not { } remaining || readAt is not { } anchor)
        {
            return null;
        }

        var elapsedAtRead = TimeSpan.FromMilliseconds(duration - remaining);
        var since = now - anchor;

        if (since < TimeSpan.Zero)
        {
            since = TimeSpan.Zero;
        }

        var position = elapsedAtRead + since;
        var length = TimeSpan.FromMilliseconds(duration);

        // Clamped rather than allowed to run past the end: the next reading is at most a few seconds
        // away, and a bar that overshoots to 103% is a more obvious lie than one that sits at the end.
        return position > length ? length : position;
    }

    /// <summary>The record's length, when the station gave one.</summary>
    public static TimeSpan? Duration(NowPlayingTrack? track) =>
        track?.DurationMs is { } ms && ms > 0 ? TimeSpan.FromMilliseconds(ms) : null;
}
